## Building DataBase for Population in Melbourne

**Reference**

Australian Bureau of Statistics. (2024). *National, state and territory population* (Catalogue No. 3101.0). 

https://www.abs.gov.au/statistics/people/population/national-state-and-territory-population

In [22]:
import pandas as pd

In [23]:
# Load the data sheet 1(Table 1)
df = pd.read_excel('./ABS Raw Datasets/ABS_Population_Estimates.xlsx', sheet_name='Table 1')

# Display the first few rows of the dtaaset
df.head(10)

Australian Bureau of Statistics       Unnamed: 1  \
0                          Regional population, 2021              NaN   
1   Released at 11.30am (Canberra time) 26 July 2022              NaN   
2  Table 1. Estimated resident population, Statis...              NaN   
3                                                NaN              NaN   
4                                                NaN              NaN   
5                                                NaN              NaN   
6                                           S/T code         S/T name   
7                                                NaN              NaN   
8                                                  1  New South Wales   
9                                                  1  New South Wales   

   Unnamed: 2   Unnamed: 3 Unnamed: 4      Unnamed: 5 Unnamed: 6  Unnamed: 7  \
0         NaN          NaN        NaN             NaN        NaN         NaN   
1         NaN          NaN        NaN             NaN        NaN         NaN   
2         NaN          NaN        NaN             NaN        NaN         NaN   
3         NaN          NaN        NaN             NaN        NaN         NaN   
4         NaN          NaN        NaN             NaN        NaN         NaN   
5         NaN          NaN        NaN             NaN        NaN         NaN   
6  GCCSA code   GCCSA name   SA4 code        SA4 name   SA3 code    SA3 name   
7         NaN          NaN        NaN             NaN        NaN         NaN   
8       1RNSW  Rest of NSW        101  Capital Region      10102  Queanbeyan   
9       1RNSW  Rest of NSW        101  Capital Region      10102  Queanbeyan   

  Unnamed: 8 Unnamed: 9  ... Unnamed: 27 Unnamed: 28 Unnamed: 29 Unnamed: 30  \
0        NaN        NaN  ...         NaN         NaN         NaN         NaN   
1        NaN        NaN  ...         NaN         NaN         NaN         NaN   
2        NaN        NaN  ...         NaN         NaN         NaN         NaN   
3        NaN        NaN  ...         NaN         NaN         NaN         NaN   
4        NaN        NaN  ...         NaN         NaN         NaN         NaN   
5        NaN        NaN  ...        2018        2019        2020        2021   
6   SA2 code   SA2 name  ...         no.         no.         no.         no.   
7        NaN        NaN  ...         NaN         NaN         NaN         NaN   
8  101021007  Braidwood  ...        4140        4211        4273        4330   
9  101021008    Karabar  ...        8507        8488        8519        8546   

  Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34 Unnamed: 35  \
0         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN   
4         NaN      Change         NaN         NaN         NaN   
5         NaN   2011-2021         NaN         NaN        Area   
6         NaN         no.           %         NaN         km2   
7         NaN         NaN         NaN         NaN         NaN   
8         NaN         872        25.2         NaN      3418.4   
9         NaN        -546          -6         NaN           7   

               Unnamed: 36  
0                      NaN  
1                      NaN  
2                      NaN  
3                      NaN  
4                      NaN  
5  Population density 2021  
6              persons/km2  
7                      NaN  
8                      1.3  
9                   1223.9  

[10 rows x 37 columns]

In [24]:
# Drop the first 5 rows
df = df.drop(range(5))

# Reset the index
df = df.reset_index(drop=True)

# Display the first few rows of the dataset
df.head(10)

Australian Bureau of Statistics       Unnamed: 1  Unnamed: 2  \
0                                         NaN              NaN         NaN   
1                                    S/T code         S/T name  GCCSA code   
2                                         NaN              NaN         NaN   
3                                           1  New South Wales       1RNSW   
4                                           1  New South Wales       1RNSW   
5                                           1  New South Wales       1RNSW   
6                                           1  New South Wales       1RNSW   
7                                           1  New South Wales       1RNSW   
8                                           1  New South Wales       1RNSW   
9                                           1  New South Wales       1RNSW   

    Unnamed: 3 Unnamed: 4      Unnamed: 5 Unnamed: 6  Unnamed: 7 Unnamed: 8  \
0          NaN        NaN             NaN        NaN         NaN        NaN   
1   GCCSA name   SA4 code        SA4 name   SA3 code    SA3 name   SA2 code   
2          NaN        NaN             NaN        NaN         NaN        NaN   
3  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021007   
4  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021008   
5  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021009   
6  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021010   
7  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021012   
8  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021610   
9  Rest of NSW        101  Capital Region      10102  Queanbeyan  101021611   

                        Unnamed: 9  ... Unnamed: 27 Unnamed: 28 Unnamed: 29  \
0                              NaN  ...        2018        2019        2020   
1                         SA2 name  ...         no.         no.         no.   
2                              NaN  ...         NaN         NaN         NaN   
3                        Braidwood  ...        4140        4211        4273   
4                          Karabar  ...        8507        8488        8519   
5                       Queanbeyan  ...       11447       11450       11437   
6                Queanbeyan - East  ...        5072        5117        5077   
7  Queanbeyan West - Jerrabomberra  ...       13008       12935       12796   
8                          Googong  ...        3453        4484        5448   
9             Queanbeyan Surrounds  ...       15919       16048       16267   

  Unnamed: 30 Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34 Unnamed: 35  \
0        2021         NaN   2011-2021         NaN         NaN        Area   
1         no.         NaN         no.           %         NaN         km2   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3        4330         NaN         872        25.2         NaN      3418.4   
4        8546         NaN        -546          -6         NaN           7   
5       11370         NaN         655         6.1         NaN         4.8   
6        5093         NaN         142         2.9         NaN          13   
7       12743         NaN        -480        -3.6         NaN        13.7   
8        6245         NaN        6194     12145.1         NaN         6.9   
9       16523         NaN        1962        13.5         NaN      3047.7   

               Unnamed: 36  
0  Population density 2021  
1              persons/km2  
2                      NaN  
3                      1.3  
4                   1223.9  
5                   2387.7  
6                    391.7  
7                    931.9  
8                    902.9  
9                      5.4  

[10 rows x 37 columns]

In [25]:
df.iloc[0:2,10:]

# Drop column 32-34
df = df.drop(columns=df.columns[31:35])

df.iloc[0:2,10:]

Unnamed: 10 Unnamed: 11 Unnamed: 12 Unnamed: 13 Unnamed: 14 Unnamed: 15  \
0        2001        2002        2003        2004        2005        2006   
1         no.         no.         no.         no.         no.         no.   

  Unnamed: 16 Unnamed: 17 Unnamed: 18 Unnamed: 19  ... Unnamed: 23  \
0        2007        2008        2009        2010  ...        2014   
1         no.         no.         no.         no.  ...         no.   

  Unnamed: 24 Unnamed: 25 Unnamed: 26 Unnamed: 27 Unnamed: 28 Unnamed: 29  \
0        2015        2016        2017        2018        2019        2020   
1         no.         no.         no.         no.         no.         no.   

  Unnamed: 30 Unnamed: 35              Unnamed: 36  
0        2021        Area  Population density 2021  
1         no.         km2              persons/km2  

[2 rows x 23 columns]

In [26]:
# Get the list of header column 10-
header_col1 = df.iloc[0, 10:]

# Get the list of header column 0-26
header_col2 = df.iloc[1, 0:10]

# Merge the two lists
header_col = pd.concat([header_col2, header_col1]).tolist()

header_col

['S/T code',
 'S/T name',
 'GCCSA code',
 'GCCSA name',
 'SA4 code',
 'SA4 name',
 'SA3 code',
 'SA3 name',
 'SA2 code',
 'SA2 name',
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018,
 2019,
 2020,
 2021,
 'Area',
 'Population density 2021']

In [27]:
# Change the header names, same as header_col

current_columns = df.columns.tolist()

# Mapping the column names
rename_dict = dict(zip(current_columns, header_col))

# rename()
df.rename(columns=rename_dict, inplace=True)

# Drop the first 2 rows
df = df.drop(range(3))
df = df.reset_index(drop=True)

# Check the first few rows of the dataset
df.head()


S/T code         S/T name GCCSA code   GCCSA name SA4 code        SA4 name  \
0        1  New South Wales      1RNSW  Rest of NSW      101  Capital Region   
1        1  New South Wales      1RNSW  Rest of NSW      101  Capital Region   
2        1  New South Wales      1RNSW  Rest of NSW      101  Capital Region   
3        1  New South Wales      1RNSW  Rest of NSW      101  Capital Region   
4        1  New South Wales      1RNSW  Rest of NSW      101  Capital Region   

  SA3 code    SA3 name   SA2 code                         SA2 name  ...  \
0    10102  Queanbeyan  101021007                        Braidwood  ...   
1    10102  Queanbeyan  101021008                          Karabar  ...   
2    10102  Queanbeyan  101021009                       Queanbeyan  ...   
3    10102  Queanbeyan  101021010                Queanbeyan - East  ...   
4    10102  Queanbeyan  101021012  Queanbeyan West - Jerrabomberra  ...   

    2014   2015   2016   2017   2018   2019   2020   2021    Area  \
0   3762   3849   3950   4039   4140   4211   4273   4330  3418.4   
1   8731   8603   8531   8526   8507   8488   8519   8546       7   
2  11199  11213  11230  11355  11447  11450  11437  11370     4.8   
3   4967   4961   4970   5013   5072   5117   5077   5093      13   
4  13193  13164  13150  13083  13008  12935  12796  12743    13.7   

  Population density 2021  
0                     1.3  
1                  1223.9  
2                  2387.7  
3                   391.7  
4                   931.9  

[5 rows x 33 columns]

In [28]:

# Filter out the data using SA4 name contains 'Melbourne'
df_melbourne = df[df['SA4 name'].str.contains('Melbourne', case=False, na=False)]
df_melbourne = df_melbourne.reset_index(drop=True)

df_melbourne.head()


S/T code  S/T name GCCSA code         GCCSA name SA4 code  \
0        2  Victoria      2GMEL  Greater Melbourne      206   
1        2  Victoria      2GMEL  Greater Melbourne      206   
2        2  Victoria      2GMEL  Greater Melbourne      206   
3        2  Victoria      2GMEL  Greater Melbourne      206   
4        2  Victoria      2GMEL  Greater Melbourne      206   

            SA4 name SA3 code            SA3 name   SA2 code  \
0  Melbourne - Inner    20601  Brunswick - Coburg  206011106   
1  Melbourne - Inner    20601  Brunswick - Coburg  206011107   
2  Melbourne - Inner    20601  Brunswick - Coburg  206011109   
3  Melbourne - Inner    20601  Brunswick - Coburg  206011495   
4  Melbourne - Inner    20601  Brunswick - Coburg  206011496   

            SA2 name  ...   2014   2015   2016   2017   2018   2019   2020  \
0     Brunswick East  ...  10439  11062  11716  12149  12382  12588  13045   
1     Brunswick West  ...  14192  14344  14523  14549  14685  14840  14991   
2  Pascoe Vale South  ...  10122  10251  10465  10693  10825  10861  10819   
3  Brunswick - North  ...  12922  13225  13581  13725  13923  14016  14116   
4  Brunswick - South  ...  13233  13574  13854  13979  14192  14326  14221   

    2021 Area Population density 2021  
0  12963  2.2                  5978.7  
1  14499  3.2                  4560.2  
2  10461    3                  3500.2  
3  13091  2.4                    5434  
4  13215  2.7                  4834.6  

[5 rows x 33 columns]

In [29]:
df_melbourne.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   S/T code                 344 non-null    object
 1   S/T name                 344 non-null    object
 2   GCCSA code               344 non-null    object
 3   GCCSA name               344 non-null    object
 4   SA4 code                 344 non-null    object
 5   SA4 name                 344 non-null    object
 6   SA3 code                 344 non-null    object
 7   SA3 name                 344 non-null    object
 8   SA2 code                 344 non-null    object
 9   SA2 name                 344 non-null    object
 10  2001                     344 non-null    object
 11  2002                     344 non-null    object
 12  2003                     344 non-null    object
 13  2004                     344 non-null    object
 14  2005                     344 non-null    o

In [30]:
# Filter out Melbourne CBD using SA2 name contains 'Melbourne CBD'
df_CBD = df_melbourne[df_melbourne['SA2 name'].str.contains('Melbourne CBD', case=False, na=False)]
df_CBD = df_CBD.reset_index(drop=True)

df_CBD.head()


S/T code  S/T name GCCSA code         GCCSA name SA4 code  \
0        2  Victoria      2GMEL  Greater Melbourne      206   
1        2  Victoria      2GMEL  Greater Melbourne      206   
2        2  Victoria      2GMEL  Greater Melbourne      206   

            SA4 name SA3 code        SA3 name   SA2 code  \
0  Melbourne - Inner    20604  Melbourne City  206041503   
1  Melbourne - Inner    20604  Melbourne City  206041504   
2  Melbourne - Inner    20604  Melbourne City  206041505   

                SA2 name  ...   2014   2015   2016   2017   2018   2019  \
0   Melbourne CBD - East  ...  10455  11049  11633  11893  12346  12569   
1  Melbourne CBD - North  ...  11297  12895  14230  15810  17150  18394   
2   Melbourne CBD - West  ...  11874  13218  14318  16896  18119  18780   

    2020   2021 Area Population density 2021  
0  12398  10205  0.8                 12724.4  
1  19192  17439  0.6                 31052.4  
2  18835  16179    1                 16096.9  

[3 rows x 33 columns]

In [31]:
df_CBD.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   S/T code                 3 non-null      object
 1   S/T name                 3 non-null      object
 2   GCCSA code               3 non-null      object
 3   GCCSA name               3 non-null      object
 4   SA4 code                 3 non-null      object
 5   SA4 name                 3 non-null      object
 6   SA3 code                 3 non-null      object
 7   SA3 name                 3 non-null      object
 8   SA2 code                 3 non-null      object
 9   SA2 name                 3 non-null      object
 10  2001                     3 non-null      object
 11  2002                     3 non-null      object
 12  2003                     3 non-null      object
 13  2004                     3 non-null      object
 14  2005                     3 non-null      objec

### Table Structure Example

**Population**

| population_id | region_name | region_code | year | population |
|:---|:---|:---|:---|:---|
|1|Melbourne CBD - East | 206041503	| 2014 | 10455 |

In [32]:
# column names
df_CBD.columns

Index([               'S/T code',                'S/T name',
                    'GCCSA code',              'GCCSA name',
                      'SA4 code',                'SA4 name',
                      'SA3 code',                'SA3 name',
                      'SA2 code',                'SA2 name',
                            2001,                      2002,
                            2003,                      2004,
                            2005,                      2006,
                            2007,                      2008,
                            2009,                      2010,
                            2011,                      2012,
                            2013,                      2014,
                            2015,                      2016,
                            2017,                      2018,
                            2019,                      2020,
                            2021,                    'Area',
       'Population densi

In [33]:
# Remain SA2 name, SA2 code, years(2001-2021)
df_CBD = df_CBD[['SA2 name', 'SA2 code', 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021]]

# Rename columns
df_CBD.rename(columns = {'SA2 name': 'region_name', 'SA2 code': 'region_code'}, inplace = True)

# Get the year list
year_list = df_CBD.columns[2:].tolist()
print(year_list)

# Pivot the table: region_name, region_code, year, population
df_CBD = df_CBD.melt(id_vars=['region_name', 'region_code'],
                     value_vars=year_list,
                     var_name='year',
                     value_name='population')

# Check the table
print(df_CBD.head())

# Check the table structure
print(df_CBD.info())

[2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021]
             region_name region_code  year population
0   Melbourne CBD - East   206041503  2001       3997
1  Melbourne CBD - North   206041504  2001       1657
2   Melbourne CBD - West   206041505  2001       1990
3   Melbourne CBD - East   206041503  2002       4794
4  Melbourne CBD - North   206041504  2002       2282
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   region_name  63 non-null     object
 1   region_code  63 non-null     object
 2   year         63 non-null     object
 3   population   63 non-null     object
dtypes: object(4)
memory usage: 2.1+ KB
None


In [34]:
# Convert Datatype
df_CBD['region_code'] = df_CBD['region_code'].astype(int)
df_CBD['year'] = df_CBD['year'].astype(int)
df_CBD['population'] = df_CBD['population'].astype(int)

# Check the table structure
print(df_CBD.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   region_name  63 non-null     object
 1   region_code  63 non-null     int64 
 2   year         63 non-null     int64 
 3   population   63 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 2.1+ KB
None


In [35]:
df_CBD.head()

region_name  region_code  year  population
0   Melbourne CBD - East    206041503  2001        3997
1  Melbourne CBD - North    206041504  2001        1657
2   Melbourne CBD - West    206041505  2001        1990
3   Melbourne CBD - East    206041503  2002        4794
4  Melbourne CBD - North    206041504  2002        2282

In [36]:
# Calculate Melbourne CBD population yearly by summing up the population of East, North, West
total_rows = []

for year in df_CBD['year'].unique():
    # Filter by years
    pop_year = df_CBD[df_CBD['year'] == year]

    # Calculate the total population
    total_pop = pop_year['population'].sum()

    # Create a new row dictionary (region_code = None)
    new_row = {
        'region_name' : 'Melbourne CBD - Total',
        'region_code' : None, # None because it is not a real region code
        'year' : year,
        'population' : total_pop
    }

    total_rows.append(new_row)

df_total_rows = pd.DataFrame(total_rows)

# Append total rows to the original table
df_CBD = pd.concat([df_CBD, df_total_rows], ignore_index=True)

# Add population_id column
df_CBD['population_id'] = range(1, len(df_CBD) + 1)

df_CBD = df_CBD[['population_id', 'region_name', 'region_code', 'year', 'population']]

# Check the table
df_CBD.tail()



population_id            region_name region_code  year  population
79             80  Melbourne CBD - Total        None  2017       44599
80             81  Melbourne CBD - Total        None  2018       47615
81             82  Melbourne CBD - Total        None  2019       49743
82             83  Melbourne CBD - Total        None  2020       50425
83             84  Melbourne CBD - Total        None  2021       43823

In [37]:
df_CBD.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   population_id  84 non-null     int64 
 1   region_name    84 non-null     object
 2   region_code    63 non-null     object
 3   year           84 non-null     int64 
 4   population     84 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 3.4+ KB


In [38]:
# Check the maximum length of region_code and region_name
max_region_name_length = df_CBD['region_name'].apply(len).max()
max_region_code_length = df_CBD['region_code'].apply(lambda x: len(str(x)) if x is not None else 0).max()

print(f"Maximum length of region_name: {max_region_name_length}")
print(f"Maximum length of region_code: {max_region_code_length}")


Maximum length of region_name: 21
Maximum length of region_code: 9


In [39]:
# Save the table
df_CBD.to_csv('population.csv', index=False)